File: node/examples/quickstart-extract.ts
/**
 * Quickstart: extract structured data from a URL.
 *
 * /v1/extract performs fast, deterministic extraction driven by your JSON Schema.
 * Annotate each property with `x-css` (a CSS selector applied to the page) or
 * `x-regex` (a pattern matched against the page text, capture group 1) to tell
 * the extractor where each field lives. Fields without a selector are best-effort.
 *
 * Run with:
 *   GHOSTCRAWL_API_KEY=gck_live_YOUR_KEY npx tsx examples/quickstart-extract.ts
 */
import { createGhostCrawlClient } from "../src/index.js";

// Selectors target the structure of https://example.com (an <h1> title and a
// descriptive <p>). Point TARGET_URL at your own page and adjust the selectors.
const PAGE_SCHEMA = {
  type: "object",
  properties: {
    title: { type: "string", "x-css": "h1" },
    description: { type: "string", "x-css": "p" },
  },
};

async function main(): Promise<void> {
  const token = process.env.GHOSTCRAWL_API_KEY;
  if (!token) {
    console.error("Set GHOSTCRAWL_API_KEY in your environment.");
    process.exit(1);
  }

  const target = process.env.TARGET_URL ?? "https://example.com";
  const client = createGhostCrawlClient({ token });

  const result = await client.extract({ url: target, schema: PAGE_SCHEMA });
  console.log("Extracted data:");
  // The API returns { url, data, token_estimate }; `data` holds the structured fields.
  console.log(JSON.stringify(result.data ?? result, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: python/examples/quickstart_extract.py
"""Quickstart: extract structured data from a page.

Run with:
    GHOSTCRAWL_API_KEY=gck_live_YOUR_KEY python examples/quickstart_extract.py
"""

from __future__ import annotations

import asyncio
import os
import sys

from ghostcrawl import GhostCrawlClient, InvalidRequestError


PRODUCT_SCHEMA = {
    "type": "object",
    "properties": {
        "title": {"type": "string"},
        "price": {"type": "number"},
        "in_stock": {"type": "boolean"},
    },
    "required": ["title"],
}


async def main() -> None:
    token = os.environ.get("GHOSTCRAWL_API_KEY")
    if not token:
        print("Set GHOSTCRAWL_API_KEY in your environment.", file=sys.stderr)
        sys.exit(1)

    target = os.environ.get("TARGET_URL", "https://example.com")

    async with GhostCrawlClient(token=token) as client:
        try:
            data = await client.extract(url=target, schema=PRODUCT_SCHEMA)
        except InvalidRequestError as exc:
            # The page may not contain every required field — the server
            # validates the extracted result against the schema and returns
            # a 422 when a required property cannot be found.
            print(f"Extraction did not satisfy the schema (HTTP {exc.status_code}).")
            if exc.body:
                print(exc.body)
            return

        print("Extracted data:")
        print(f"  title: {data.get('title')}")
        print(f"  price: {data.get('price')}")
        print(f"  in_stock: {data.get('in_stock')}")


if __name__ == "__main__":
    asyncio.run(main())
